docs/world-compat-details/index.md

```diff
@@ -365,4 +365,334 @@ LBT 扩展指令集，`magic` = `0x42540001`
+
+glibc 提供了两个函数，可以被用于注册信号处理函数，这两个函数分别是 `sigaction` 和 `signal`。
+其中，`signal` 函数是对 `sigaction` 函数的封装。这样，无论是如何注册的信号处理函数，
+都会在第三个参数中接受到一个 `ucontext_t` 结构体指针。但是根据 glibc [文档](https://man7.org/linux/man-pages/man2/signal.2.html)，
+`signal` 函数注册的信号处理函数只应该接受第一个参数。这意味着，如果 `signal` 函数注册的信号处理函数
+遵循了文档的要求，只接受第一个参数，那么它会忽略第三个参数，即`ucontext_t` 结构体指针会被忽略。
+这意味着，`signal` 函数注册的信号处理函数，是新旧世界无关的，不需要针对这样的信号处理函数做任何兼容性处理。
+而 `sigaction` 函数注册的信号处理函数，是新旧世界相关的，需要额外的兼容性处理。一种可能的兼容性处理是，
+当用户调用 `sigaction` 函数注册一个接受旧世界的 `ucontext_t` 结构体信号处理函数时，实际不注册这个函数，
+而是注册一个包裹函数，这个包裹函数接受新世界的 `ucontext_t` 结构体，并在栈上相应构造一个旧世界的 `ucontext_t` 结构体，
+然后调用用户提供的信号处理函数；在用户提供的信号处理函数返回后，将旧世界的 `ucontext_t` 结构体的内容拷贝回新世界的 `ucontext_t` 结构体中。
+需要注意的是，信号处理函数无法接受自定义的参数，这意味着原始用户提供的信号处理函数的地址必须以某种方式保存下来，
+以便包裹函数调用它。而在实现这一点时要十分小心地处理锁和信号屏蔽，因为整个这个注册信号处理函数的过程不再是原子的。
+
+无论如何处理，都不会改变新旧世界 `ucontext_t` 结构体完全不兼容的事实，也无法扭转旧世界 glibc
+和旧世界内核 `ucontext_t` 结构体存在差异的现状，因此如果出现了新旧世界可执行程序和动态连接库的混合连接，
+并且存在将 `ucontext_t` 结构体在新旧世界间传递的情况，始终无法保证正确识别 `ucontext_t` 结构体是新世界的还是旧世界的。
+不过幸运的事，这样的情况十分罕见。
+
+下列函数涉及 `ucontext_t` 结构体，因此新旧世界的函数完全不二进制兼容：
+
+- `getcontext`
+- `setcontext`
+- `makecontext`
+- `swapcontext`
+
+下列函数涉及注册能接收 `ucontext_t` 结构体的信号处理函数，因此需要额外的兼容性处理：
+
+- `sigaction`
+
+下列函数注册的信号处理函数会忽略 `ucontext_t` 结构体指针，因此无需特殊处理：
+
+- `signal`
+
+#### longjmp 相关
+
+`setjmp`、`sigsetjmp`、`longjmp`、`siglongjmp` 函数是用于实现非局部跳转的函数。
+其中 `set*jmp` 函数会将当前的程序状态（可选地，包括当前的信号掩码，即 sigmask）保存到一个 `jmp_buf` 结构体中，
+而 `*longjmp` 函数则将上述状态恢复。这些函数的行为与 `*context` 函数类似。在新旧世界中，
+`jmp_buf` 结构体的定义完全一致，因此这些 `*jmp` 函数是完全二进制兼容的。
+
+下列函数涉及 `jmp_buf` 结构体，在新旧世界间完全兼容：
+
+- `setjmp`
+- `sigsetjmp`
+- `longjmp`
+- `siglongjmp`
+
+#### `fstat` 相关
+
+新世界的内核中，缺少 `fstat` 和 `newfstatat` 系统调用。这两个系统调用可以被用于获取文件的元数据。
+其中，`fstat` 可以获取一个打开的文件描述符 （file descriptor, fd）所对应的文件的元数据；
+而 `newfstatat` 则既可以获取一个打开的文件描述符所对应的文件的元数据，也可以获取一个路径所对应的文件的元数据。
+取代了这两个系统调用的是 `statx`。`statx` 的功能与 `newfstatat` 的功能基本一致，
+但是可以按需提供更多的信息。
+
+在 2.38 的 glibc 中，所有 `*stat*` 函数会在编译期检查内核是否提供了 `fstat` 或 `newfstatat`，
+如果没有，那么这些函数会调用 `statx`[^7]。这意味着，与旧世界相比，这些 `*stat*` 函数对外呈现的行为是不变的，
+新旧世界的函数是二进制兼容的。本节会着重讨论其向内核发出系统调用的行为区别。
+
+[^7]: 这些函数最终会调用 [__fstatat64_time64](https://elixir.bootlin.com/glibc/glibc-2.38/source/sysdeps/unix/sysv/linux/fstatat64.c#L157)
+
+对于旧世界上基于 Chromium 的浏览器和 Electron 的应用程序而言，其基于 seccomp 的沙箱机制会针对 `statx`
+[返回](https://chromium.googlesource.com/chromium/src/sandbox/+/7462a4fd179376882292be2381a22df6819041c7%5E%21)
+`ENOSYS` 错误，期待其调用 `fstat` 或 `newfstatat`。
+这是因为，seccomp [无法审查](https://lwn.net/Articles/799557/)系统调用中的指针参数。
+而 Chromium 的一种沙箱规则则要求程序只能操作已经打开的 fd，而不能访问任何系统路径，
+因此只能放行（在新世界龙架构内核中不存在的）`fstat` ，并通过 `SIGSYS` 的钩子[检查](https://chromium.googlesource.com/chromium/src/sandbox/+/b3267c8b40b6133b2db5475caed8f6722837a95e%5E%21/#F2) `newfstatat` 并将其重写为 `fstat`。
+
+为了能让这部分程序正常运行，需要调整上述函数的行为，当 `statx` 返回 `ENOSYS` 时，
+改为使用 `fstat` 或 `newfstatat`；同时，需要在新世界的内核中补充 `fstat` 和 `newfstatat` 的实现。
+
+glibc 中下列导出函数涉及 `fstat` 和 `newfstatat`，为兼容 Chromium 的沙箱机制，需要额外的兼容性处理：
```

Comment:
As it's possible for the Chromium sandbox logic to change later, let's be clearer by stating we're focusing on the unadapted version of it.

```suggestion
glibc 中下列导出函数涉及 `fstat` 和 `newfstatat`，为兼容未经适配的 Chromium 的沙箱机制，需要额外的兼容性处理：
```